src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -3955,7 +3955,9 @@ public function set_attribute( $name, $value ): bool {
 			 *
 			 * @see https://html.spec.whatwg.org/#attributes-3
 			 */
-			$escaped_new_value = in_array( $comparable_name, wp_kses_uri_attributes(), true ) ? esc_url( $value ) : esc_attr( $value );
+			$escaped_new_value = in_array( $comparable_name, wp_kses_uri_attributes(), true )
+				? esc_url( $value )
```

Comment:
The `esc_url()` function also does some stuff with entities, namely:

https://github.com/WordPress/wordpress-develop/blob/fe2b33f27b8eef2b42771414a9464a3508cbca3a/src/wp-includes/formatting.php#L4522-L4527

To avoid that, I think using `esc_url_raw()` is preferable:

```suggestion
				? htmlspecialchars( esc_url_raw( $value ) )
```

Reply:
This line is only a formatting change, I broke the ternary onto three lines.

Can we consider that suggestion in a separate ticket and PR?

Reply:
Ah sure, but it it is related to the double-escaping issue, because `esc_url()` does `wp_kses_normalize_entities()` which includes converting all `&` to `&amp;` among other things. For example, these:

```php
var_dump( esc_url( 'http://example.com/?foo&amp;bar' ) );
var_dump( esc_url( 'http://example.com/?foo&#038;bar' ) );
var_dump( esc_url( 'http://example.com/?foo&bar' ) );
```

All [result in](https://playground.wordpress.net/php-playground.html#eyJjb2RlIjoiPD9waHBcbmhlYWRlcignY29udGVudC10eXBlOiB0ZXh0L3BsYWluJyk7XG5cbi8vIFdvcmRQcmVzcyBpcyBhdmFpbGFibGUgaWYgeW91IG5lZWQgaXQhXG5yZXF1aXJlICcvd29yZHByZXNzL3dwLWxvYWQucGhwJztcbnZhcl9kdW1wKCBlc2NfdXJsKCAnaHR0cDovL2V4YW1wbGUuY29tLz9mb28mYW1wO2JhcicgKSApO1xudmFyX2R1bXAoIGVzY191cmwoICdodHRwOi8vZXhhbXBsZS5jb20vP2ZvbyYjMDM4O2JhcicgKSApO1xudmFyX2R1bXAoIGVzY191cmwoICdodHRwOi8vZXhhbXBsZS5jb20vP2ZvbyZiYXInICkgKTtcbiIsInBocCI6IjguNCIsIndwIjoiNi44In0=):

```php
string(32) "http://example.com/?foo&#038;bar"
string(32) "http://example.com/?foo&#038;bar"
string(32) "http://example.com/?foo&#038;bar"
```

I don't really see why a URI attribute should be treated differently at the HTML API layer. So this could instead just be:

```
$escaped_new_value = strtr(
	$value,
	array(
		'<' => '&lt;',
		'>' => '&gt;',
		'&' => '&amp;',
		'"' => '&quot;',
		"'" => '&apos;',
	)
);
```

Reply:
You're absolutely right that the double-escaping issue also exists in `esc_url()`, that's unfortunate. `esc_url()` does a lot of things, it doesn't seem as easily replaced as `esc_attr()` or `esc_html()`. I suspect more discussion will be necessary regarding URL-related changes which is why I'm reluctant to include those changes here.

See r58472 and r58473.

Reply:
OK, sounds good.